src/outbox/outboxActions.js

```diff
@@ -46,23 +47,57 @@ export const messageSendComplete = (localMessageId: number): Action => ({
   local_message_id: localMessageId,
 });
 
-export const trySendMessages = (dispatch: Dispatch, getState: GetState): boolean => {
+export const trySendMessages = async (dispatch: Dispatch, getState: GetState): Promise<boolean> => {
   const state = getState();
   const auth = getAuth(state);
   const outboxToSend = state.outbox.filter(outbox => !outbox.isSent);
+  const oneWeekAgo = Date.now() / 1000 - 60 * 60 * 24 * 7;
   try {
-    outboxToSend.forEach(async item => {
-      await api.sendMessage(
-        auth,
-        item.type,
-        isPrivateOrGroupNarrow(item.narrow) ? item.narrow[0].operand : item.display_recipient,
-        item.subject,
-        item.markdownContent,
-        item.timestamp,
-        state.session.eventQueueId,
-      );
+    for (const item of outboxToSend) {
+      // If a message has spent over a week in the outbox, it's probably too
+      // stale to try sending it.
+      //
+      // TODO: instead of just throwing these away, create an "unsendable" state
+      // (including a reason for unsendability), and transition old messages to
+      // that instead.
+      if (item.timestamp < oneWeekAgo) {
+        dispatch(deleteOutboxMessage(item.id));
+        continue;
+      }
+
+      const to = ((): string => {
+        const { narrow } = item;
+        // TODO: can this test be `if (item.type === private)`?
+        if (isPrivateOrGroupNarrow(narrow)) {
```

Comment:
Yeah, it's kind of odd that this code (already in the status quo) uses a mix of `narrow` and `display_recipient`.

I feel like `narrow` shouldn't be necessary -- that information should be covered by `display_recipient`. Besides which, `narrow` is kind of a confused thing to use here, because a narrow doesn't necessarily correspond to any one conversation.

Reply:
(This doesn't call for any code change in this PR; just replying to the thought in the TODO comment.)

Reply:
I think the *right* thing to do here is to split private messages from stream messages in the type system, and provide a canonical discriminator method.

I tried to do that at one point, only to discover that it's going to require a later version of Flow than we have. (v0.100, I think? One of the versions with object-spread improvements, anyway.)

Reply:
> I think the _right_ thing to do here is to split private messages from stream messages in the type system, and provide a canonical discriminator method.

That would certainly also be good!